cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+priority mechanism.
+
+#### Right-hand Sides: Rewrite Expressions
+
+Given a rule whose pattern has matched, we now need to compute the
+rewritten term that replaces the original input term. This rewrite is
+specified by the right-hand side (RHS), which consists of an
+*expression* that generates a new term. This expression can use parts
+of the input term that have been captured by variables in the
+pattern.
+
+We have already seen a few examples of this above: simple term
+expressions, with variables used in place of concrete subterms where
+desired. A typical term-rewrite system allows just a few options in
+the output expression:
+
+* Terms, with sub-expressions as arguments;
+* Constant primitives (`42` or `$Symbol`); and
+* Captured variable values (`x`).
+
+The options are more limited in expressions than in patterns (e.g.,
+there are no wildcards) because a pattern is matching on a range of
+possible terms while an expression must specify a specific rewrite
+result.
+
+### Rewrite Steps and Intermediate Terms
+
+Now that we can specify rewrites via a list of rules, we can study how
+the top-level execution of a term-rewriting system proceeds. Much of
+the power of term-rewriting comes from the fact that rewrites can
+*chain together* into a multi-step traversal through several
+intermediate terms before the final answer is computed.
+
+For a simple example, consider the following rules:
+
+```plain
+    (A (B x)) -> (C x)
+    (C (D x)) -> (E x)
+    (C (F x)) -> (G x)
+```
+
+This set of rules will rewrite `(A (B (D 42)))` to `(C (D 42))`, then
+to `(E 42)` (via the first and third rules respectively).
+
+How is this useful? First, rewriting one term to another (here, `C` at
+the top level) that in turn appears in the left-hand side of other
+rules allows us to *factor* a "program" of term-rewriting rules in the
+same way that imperative programs are factored into separate
+functions.[^4] The usual advantages of a well-factored program, where
+each problem is solved with a small step that "reduces to a previously
+solved problem", apply here.
+
+Second, repeating the rewrite step is actually what grants
+term-rewriting its Turing-completeness: it allows for arbitrary
+control flow.[^5] This might be useful in cases where, for example, a
+term-rewriting program needs "loop" over a list of elements in the
+input: it can recurse and use intermediate terms to store state.
+
+While this full generality may not be used often in the
+domain-specific applications of term-rewriting that emphasize its
+pattern-matching (such as instruction selectors), the user should not
+be afraid to define and use intermediate terms -- rewriting into them,
+then defining additional rules to rewrite further -- when it helps to
+factor common behavior out of multiple rules, or aids in conceptual
+clarity.
+
+[^4]: In fact, ISLE actually compiles rules for different top-level
+      pattern terms (`(A ...)` and `(C ...)` in the example) into
```

Comment:
This is cool!

Does this mean `(A 1 _) -> ...` and `(A 2 _) -> ...` would share the function created for `(A ...)`?

Reply:
Yes, exactly; all of the logic to rewrite `(A ...)` is invoked from one function entry point.